v22.1/row-level-ttl.md

```diff
@@ -0,0 +1,342 @@
+
+TTLs are defined on a per-table basis using SQL statements. The syntax for creating a table with an automatically managed TTL extends the [`storage_parameter` syntax](sql-grammar.html#opt_with_storage_parameter_list). For example, the SQL statement
+
+{% include_cached copy-clipboard.html %}
+~~~ sql
+CREATE TABLE ttl_test (
+  id UUID PRIMARY KEY default gen_random_uuid(),
+  description TEXT,
+  inserted_at TIMESTAMP default current_timestamp()
+) WITH (ttl_expire_after = '3 minutes');
+~~~
+
+has the following effects:
+
+<a name="crdb-internal-expiration"></a>
+
+1. Creates a repeating [scheduled job](#view-ttl-jobs) for the `ttl_test` table.
+2. Adds a hidden column `crdb_internal_expiration` of type [`TIMESTAMPTZ`](timestamp.html) to represent the TTL.
+3. Implicitly adds the `ttl` and `ttl_automatic_column` [storage parameters](#ttl-storage-parameters).
+
+To see the hidden column and the storage parameters, enter the [`SHOW CREATE TABLE`](show-create.html) statement:
+
+{% include_cached copy-clipboard.html %}
+~~~ sql
+SHOW CREATE TABLE ttl_test;
+~~~
+
+~~~
+  table_name |                                                                                           create_statement
+-------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+  ttl_test   | CREATE TABLE public.ttl_test (
+             |     id UUID NOT NULL DEFAULT gen_random_uuid(),
+             |     description STRING NULL,
+             |     inserted_at TIMESTAMP NULL DEFAULT current_timestamp():::TIMESTAMP,
+             |     crdb_internal_expiration TIMESTAMPTZ NOT VISIBLE NOT NULL DEFAULT current_timestamp():::TIMESTAMPTZ + '00:03:00':::INTERVAL ON UPDATE current_timestamp():::TIMESTAMPTZ + '00:03:00':::INTERVAL,
+             |     CONSTRAINT ttl_test_pkey PRIMARY KEY (id ASC)
+             | ) WITH (ttl = 'on', ttl_automatic_column = 'on', ttl_expire_after = '00:03:00':::INTERVAL)
+(1 row)
+~~~
+
+To view the scheduled job, use [`SHOW SCHEDULES`](show-schedules.html):
+
+{% include_cached copy-clipboard.html %}
+~~~ sql
+SHOW SCHEDULES;
+~~~
+
+~~~
+          id         |        label         | schedule_status |        next_run        |  state  | recurrence | jobsrunning | owner |            created            |     command
+---------------------+----------------------+-----------------+------------------------+---------+------------+-------------+-------+-------------------------------+-------------------
+  747608117920104449 | sql-stats-compaction | ACTIVE          | 2022-03-25 16:00:00+00 | pending | @hourly    |           0 | node  | 2022-03-25 15:31:31.444067+00 | {}
+  747609229470433281 | row-level-ttl-112    | ACTIVE          | 2022-03-25 16:00:00+00 | NULL    | @hourly    |           0 | root  | 2022-03-25 15:37:10.613056+00 | {"tableId": 112}
+(2 rows)
+~~~
+
+## TTL storage parameters
```

Comment:
is it worth also mentioning the metrics one can use to monitor the effectiveness of the TTL? https://github.com/cockroachdb/cockroach/blob/b9a675266303528d71b388ac35dfb87f145cc7a8/pkg/sql/ttl/ttljob/ttljob.go#L167-L244 contains a list of all metrics.

Note the metrics by default are aggregated. If the user wishes to have metrics broken down by table, if `set cluster setting server.child_metrics.enabled = true;` is set AND the table has the storage parameter (ttl_label_metrics=True) (this is necessary to reduce cardinality of the metrics).

Reply:
Yes I think it's def worth mentioning! Added a table with the metrics and those notes, PTAL

Would these metrics be visible in the UI using the [advanced debug page](https://www.cockroachlabs.com/docs/v21.2/ui-debug-pages)?  

Or is this the type of thing that would be piped out to Prometheus a la https://www.cockroachlabs.com/docs/v21.2/monitor-cockroachdb-with-prometheus.html  so users should expect to get at them that way?

(Sorry, I don't know much about CRDB metrics things)

Reply:
they are both! the advanced debug page shows all these stats, and prometheus scrapes these stats from the same data source (from the `/_status/vars` endpoint on the db console)